File: builder/Builder.cs
using System.Collections.Generic;

class Ski {

    List<string> components = new List<string>();

    public void Add(string component) => components.Add(component);
}

interface Builder {
    void AddTipProtector();
    void AddEdges();
    void AddBindings();
    void AddBreaks();
}

class SkiBuilder : Builder {
    private Ski _finalProduct = new Ski();
    public void AddTipProtector() => this._finalProduct.Add("Tip Protector");
    public void AddEdges() => this._finalProduct.Add("Edges");
    public void AddBindings() => this._finalProduct.Add("Bindings");
    public void AddBreaks() => this._finalProduct.Add("Breaks");
    public Ski GetInstance(){
        Ski instance = this._finalProduct;
        Reset();
        return instance;
    }
    private void Reset() => _finalProduct = new Ski();
}

class Client{
    SkiBuilder builder = new SkiBuilder();
    Ski newSki = builder.AddBreks().AddBindings().AddEdges().GetInstance();
}

File: builder/Builder.ts
namespace patterns.Builder {
  interface Builder {
    AddTipProtector(): Builder;
    AddEdges(): Builder;
    AddBindings(): Builder;
    AddBreaks(): Builder;
  }

  class Ski {
    public parts: string[] = [];

    public listParts(): void {
      console.log(`Product parts: ${this.parts.join(", ")}\n`);
    }
  }

  class SkiBuilder implements Builder {
    private _finalProduct: Ski;

    constructor() {
      this.reset();
    }

    AddTipProtector(): SkiBuilder {
      this._finalProduct.parts.push("Tip Protector");
      return this;
    }
    AddEdges(): SkiBuilder {
      this._finalProduct.parts.push("Edges");
      return this;
    }
    AddBindings(): SkiBuilder {
      this._finalProduct.parts.push("Bindings");
      return this;
    }
    AddBreaks(): SkiBuilder {
      this._finalProduct.parts.push("Breaks");
      return this;
    }

    public GetInstance(): Ski {
      let result = this._finalProduct;
      this.reset();
      return result;
    }

    private reset(): void {
      this._finalProduct = new Ski();
    }
  }

  function Client(): void {
    let skiBuilder = new SkiBuilder();
    let concreteSki: Ski = skiBuilder.AddBindings().AddBreaks().GetInstance();
  }
}
